### DISCUSSION TITLE
[SOLVED] How to pass permanent rootflags to snapshots' kernel?

### DISCUSSION THREAD

**Discussion by FGD-Garuda** [Antynea/grub-btrfs#129] [Q&A] [answered]
Hello,

What is the right way to pass permanent rootflags to the snapshots?

So far I have in my `/etc/default/grub` config this line:

`GRUB_CMDLINE_LINUX_DEFAULT="rw quiet splash showopts rootflags=noatime"`

Then in `/boot/grub/grub-btrfs.cfg` I have:

`linux "/@snapshots/daily/root2020-12-27_19H05/boot/vmlinuz-5.8.0-33-generic" root=/dev/sda1 rw  rw quiet splash showopts rootflags=noatime rootflags=subvol="@snapshots/daily/root2020-12-27_19H05"`

Don't mind the duplication of `rw`, this one is passed fine once.
However you can see 2 `rootflags=`, one with my mount option and the other with the subvol coming from `/etc/grub.d/41_snapshot-btrfs`:

`linux \"${boot_dir_root_grub}/"${k}"\" root="${LINUX_ROOT_DEVICE}" rw ${kernel_parameters} rootflags=subvol=\""${snap_dir_name}"\""`

That above line explains the duplication of `rw` but also the addition of a second `rootflags=`.

When I boot from that snapshot, `findmnt -l` tells me the snapshot has been mounted with default `relatime`. Which means having twice `rootflags=` does not work, so then which way can I pass permanent rootflags to the snapshots?

Should I simply edit `41_snapshots-btrfs` and add `,noatime` after `rootflags=subvol=\""${snap_dir_name}"\""`?
And remove the `rw` at the same time.

That would probably work but I'm not sure it's the cleanest way of doing it. That would mean some parms can be written in the `GRUB_CMDLINE_LINUX_DEFAULT` line but some others can't and you have to go elsewhere to do it. Maybe that's how it is, maybe there is a good valid reason for that too, I don't know, that's why I'm asking. :)

Thanks

**Comment by Antynea** [comment, marked as answer]
Hello,
> That would probably work but I'm not sure it's the cleanest way of doing it. That would mean some parms can be written in the GRUB_CMDLINE_LINUX_DEFAULT line but some others can't and you have to go elsewhere to do it. Maybe that's how it is, maybe there is a good valid reason for that too, I don't know, that's why I'm asking. :)

This is a weakness of the script, which needs to be improved.

The duplication is normal, It is always the last parameter that is used for kernel parameters.
The parameter "rw" is a residue of an old version of the script.
`rootflags=subvol=\""${snap_dir_name}"\""` was placed at the end, because some users already specified this parameter in their `/etc/default/grub` configuration file and this was a problem.

While waiting for this part of the script to be rewritten, you can proceed as you described:
Edit `41_snapshots-btrfs`
remove `rw `parameter, 
modified the `rootflags` parameter by adding your options , this should be correct (not tested)
`rootflags=subvol=\""${snap_dir_name}"\"",noatime`

**Reply by FGD-Garuda** [reply]
Ok great! All that is super clear now.

Thanks Antynea and Happy New Year!

**Reply by Antynea** [reply]
Happy New Year and thank you.
I am thinking of turning this discussion into an issue, to think about correcting this problem.

**Reply by Kr1ss-XD** [reply]
> [...], which needs to be improved.

Second this, it would be **great** if `grub-btrfs` would support custom rootflags @Antynea  :exclamation: 

My current workaround is a pacman-hook patching `41_snapshots-btrfs` everytime `grub-btrfs` is updated - in fact for the reason that I'm adding `noatime` (and `compress=zstd`). That feels hacky and bugs me; often the patch needs to be adapted to work.

Maybe a new config variable could be introduced - let's say e.g. `GRUB_BTRFS_ROOTFLAGS`, and concatenate this with the `subvol=...` flag in `41_snapshots-btrfs` ?  I'd be willing to implement and test this, and submit a PR if you're interested.

**Reply by Antynea** [reply]
in addition with a custom variable, I would like the script to automatically retrieve the mount options present in the /etc/fstab file of the snapshot.

**Reply by Kr1ss-XD** [reply]
That sounds nice, ~_but_  there are no boot parameters or mount options for `/` defined in `fstab`, unless I'm missing something.~

~Basically, it contains mount rules for partitions/subvolumes _other than_ `/`. (If there's an fstab entry for `/`, it doesn't have any effect.) Hence I'd say it would be confusing if users could define rootflags at this place...~

**EDIT/** Never mind, seems like if a `/` mountpoint is defined in that file, the filesystem root _will_ be remounted with these options during the boot process. My bad.

**Reply by Antynea** [reply]
> That sounds nice, but there are no boot parameters or mount options for / defined in fstab, unless I'm missing something.

Systemd reads the /etc/fstab file.
if the line for / is present and match , it remount / with the indicated options.
```
cat /etc/fstab

UUID=bbad010b-a5d4-41a5-968f-ff9a609e93e3       /               btrfs           rw,relatime,space_cache,subvolid=257,subvol=/@racine     0 0
```
```
findmnt

/                                     /dev/sda1[/@racine]    btrfs      rw,relatime,space_cache,subvolid=257,subvol=/@racine
```

Okay, you edited your post during my reply :)

**Reply by Antynea** [reply]
So I'd like to retrieve the mount options in /etc/fstab and apply them to the snapshot. Then also allow an option to customize this variable.

**Reply by Kr1ss-XD** [reply]
Yeah that's cool. :+1: I think I could implement that if you like.

**Reply by Antynea** [reply]
treat yourself :)

**Reply by bam80** [reply]
What is your option if you want `noatime` not only for a particular snapshot, but globally?
Editing `fstab` for every mount seems hacky.
